### ISSUE TITLE
Add extraopts attribute to GaudiRun application

### ISSUE THREAD

**Issue by egede** [ganga-devs/ganga#640]
Created as a separate issue from #446.

@rob-c I would really like to keep the `extraopts` as a string. If you want to test a job running on just 1000 events, it is very easy to add this in as a string, rather than having to write a file with a single line. Now behind the scene, `Ganga` will of course have to write it into a file. To make it work that you can split a job on the `exraopts` and indeed the `options` attribute will require a bit of cleverness in order to be as efficient as possible while uploading files to `Dirac`.

**Comment by rob-c** [comment]
@egede OK I can see some sense in having it as a string (I'm keen to encourage adoption rather than fix problems in older apps).
But I'm beginning to think in this case it's a good idea to make the myOpts able to accept a list in this case or have another attribute as I'm expecting it would be very useful to break out opts files which are getting complex into smaller files which can be mixed/matched rather than something too complex.

Moving back to having a unique file per subjob loses some performance gains of GaudiRun from a job submission perspective so I'd be tempted to pickle the data into something like a reduced base64 and then pass it as a job option if it's short enough that people are typing it on the command line in Ganga.

**Comment by egede** [comment]
@rob-c The encoding of the extraopts (or similar) in the JDL for the Dirac job is something that might cause us trouble later. Another option would be that you simply upload a single file to DIRAC that contains *all* these small files for *all* the subjobs. Maybe place them in subdirs named by the subjob number. Then on the worker node, you download the complete thing but only unpack the relevant subdir. There will be an overhead in terms of downloading too many files, but I doubt that will ever be a problem.

**Comment by rob-c** [comment]
@egede There are different options I'm considering. I'm just keen to avoid spamming DIRAC with hundreds of small files which aren't much different like we're currently doing during job submission.

I am considering doing this with 1 additional LFN as you suggest but constructing it in a purely pythonic way could be tricky unless I build in a mechanism where each subjob during the prepare step to append to a global archive file and then we submit this once to DIRAC with the first subjob.

I've no idea how to cleanly hook into the submit step though and we don't always know the splitting from the master_prepare step so can't always anticipate what is to be done here.

(This also required thread-locking but that is a minor aside)

I think we don't strictly know the splitting from the RTHandler level on a per-subjob basis so there are advantages and disadvantages to each approach.

**Comment by egede** [comment]
@rob-c I believe that we create all the subjobs before we start to submit them. Would the `master_submit` method of the `Dirac` backend be the place to implement this?

**Comment by rob-c** [comment]
Possibly but I'm wondering if I should place the LFN which is uploaded in the backend instance or the Application instance in this case?

It's easy enough to submit 1 file and store it in the `job.master.application`.

I'm beginning to wonder, should I store a list of files I want uploaded in the master application and rely on the master_submit code processing a list of file objects stored in the `job.master.application` and uploading them to the grid before submitting the subjobs?

I could modify the schema entry used at the moment to be a list of objects and then only jobs which successfully prepare upload files to the grid in 1 well defined step.

**Comment by egede** [comment]
@rob-c I guess you want to persist the LFN where it got uploaded such that you can resubmit a subjob later. However, it is really not a property of the application so I would not think that is the correct place to store the information. It might require a bit more discussion in the weekly meeting where to place it (the splitter might actually be a possibility)..

**Comment by KonstantinSchubert** [comment]
I assume this issue is talking about what is now called GaudiExec? I've been trying to submit one, but it fails because:

> GangaAttributeError: GangaAttributeError:  ("Can't assign 'extraopts' as it does NOT appear in the object schema for class 'GaudiExec'",)

I know a few people how have grown very dependent on the extraopts. For example, this allows us to dynamically(!) set the database tags without modifying an options file.

```
    j.application.extraopts =  """
DaVinci().CondDBtag = '{CondDB}'
DaVinci().DDDBtag = '{DDDB}'
        """.format(**jobsDict)
```

I really would not know any other way of doing this.

**Comment by rob-c** [comment]
@KonstantinSchubert Can you try using `extraOpts` and let me know if it works for you: https://github.com/ganga-devs/ganga/blob/develop/python/GangaLHCb/Lib/Applications/GaudiExec.py#L102

**Comment by KonstantinSchubert** [comment]
@rob-c `extraOpts` seems to work.
So the issue on my side was just the spelling? `extraOpts` vs `extraopts`? Sorry if I completely missed the point of this issue.

**Comment by rob-c** [comment]
@KonstantinSchubert This issue should have been closed a while ago but I've quite a few things on my plate atm so closing this slipped past me
